### ISSUE TITLE
TC: Design for unified datasets

### ISSUE THREAD

**Issue by pcattori** [Datatamer/tamr-client#372]
# 💬 RFC

Now that we have projects in `tamr_client`, the next logical step is to add support for unified datasets.

Proposal: add a new module, `tc.datasets.unified`, with unified dataset features:
- get unified dataset from project
- get name of unified dataset
- commit unified dataset
- ...more?

## 🔦 Context

I write a lot of functions that accept a mastering project as input and then need to cobble together names of derived datasets based on the unified dataset name.

## 💻 Examples

using `tamr_unify_client`:

```python
def _compile_metrics(project: MasteringProject) -> Metrics:
    ud = project.unified_dataset()

    # keep in-memory lookup of group -> true cluster ID
    grouped_entities = project.client.datasets.by_name(
        f"{ud.name}_dedup_grouped_entities"
    )
    ...
```

using `tamr_client`:

```python
def _compile_metrics(s: tc.Session, project: tc.mastering.Project) -> Metrics:
    ud = tc.datasets.unified.from_project(s, project)

    # keep in-memory lookup of group -> true cluster ID
    instance = project.url.instance
    grouped_entities = tc.datasets.from_name(
        s,
        instance, 
        f"{ud.name}_dedup_grouped_entities"
    )
    ...
```
```

**Comment by nbateshaus** [comment]
A challenge for users of this structure is that the association between projects and unified datasets has to live in the user's head; if a user is looking at a project, they have no idea that a unified dataset is a thing that might be associated with it.

**Comment by nbateshaus** [comment]
There are many derived datasets; a unified dataset is one example. Derived dataset have a lot in common - in particular, they cannot be updated, their schema cannot be modified, and they can be refreshed. I would prefer if a derived dataset were a thing, and a unified dataset were a way to identify a particular derived dataset associated with a project.

**Comment by ianbakst** [comment]
I am inclined to agree with the refined structure that @nbateshaus just described. That might be a good way to group datasets. Would it make sense for the user could be able to get a Unified Dataset either: 1- from a project (if they do understand the project-UD relationship), or 2- by_name (if they don't know that relationship or don't care about any other aspects of the project)? Either one would return the UD as a derived dataset. ...I might be missing the point, though.

**Comment by ianbakst** [comment]
At least from my (FE) perspective, the duality of a dataset, especially a derived dataset, as being both a standalone item and associated with (or even a direct output of) a project is very real in my head.

**Comment by pcattori** [comment]
> ...able to get a Unified Dataset either: 1- from a project (if they do understand the project-UD relationship), or 2- by_name (if they don't know that relationship or don't care about any other aspects of the project)? Either one would return the UD as a derived dataset...

I like this approach but have two points of hesitation:

1. If getting the dataset `by_name`, how could we figure out that this is a derived dataset?
2. Are there additional "types" of datasets beyond "source" and "derived"? Are the "types" of datasets mutually exclusive (yes for "source" / "derived", but maybe not for other types I'm not aware of)?

---

Expanding on these:

Is there something in the JSON blob for the dataset to indicate that?Is checking for `upstream_datasets` sufficient? Or would we have to crawl some sort of DAG of dataset dependencies or recipe metadata to string along the necessary info to determine if a dataset is "derived" or not. 

@nbateshaus definitionally, are all datasets in Tamr "derived" if they are not source datasets? Or are those two ("source" and "derived") not exhaustive for the types of datasets we have?

**Comment by ianbakst** [comment]
The `upstream_datasets` is indeed right in the dataset JSON

**Comment by pcattori** [comment]
Chatted with @gbeskales and @nbateshaus about this today. The API doesn't explicitly model source vs derived datasets[1], and its not clear that "source vs derived" is a meaningful distinction for users of the Python Client.

The API *does* model unified datasets specially (there are API endpoints with `unifiedDataset` in the path), which I think is an argument for modeling unified datasets as a thing in the Python Client.

@nbateshaus ^ if you have any color to add, please do

[1]: though you *can* implicitly differentiate them based on the presence of `upstreamDatasets` in the JSON blob for the dataset resource.

**Comment by pcattori** [comment]
Resolved by #389
